### PR TITLE
cli: Initial patch for Glusterd2 CLI

### PR DESCRIPTION
**Pull request by aravindavk**
Updates: #205
Signed-off-by: Aravinda VK <mail@aravindavk.in>

### CODE REVIEW

**Comment by aravindavk** [comment]
Still WIP.

**Comment by aravindavk** [comment]
Implemented Peer probe and peer detach. All others(peer status, pool list, volume ops) are just place holder now.

Looks like Peer detach API accepts ID instead of fqdn, that is not taken care.
Cleanup required in Error and success messages.
Makefile changes are remaining for cli dependency handling and other checks

Note: This depends on restclient patch.

File: cli/cmd/peer.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+var poolCmd = &cobra.Command{
+	Use:   "pool",
+	Short: helpPeerCmd,
+}
+
+var peerProbeCmd = &cobra.Command{
+	Use:   "probe <HOSTNAME>",
+	Short: helpPeerCmd,
```
Comment (kshlm) [inline cli/cmd/peer.go]: Should be `helpPeerProbeCmd`.

File: cli/cmd/root.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+var (
+	flagXMLOutput  bool
+	flagJSONOutput bool
+)
+
+func init() {
+	initRESTClient()
+	initLog("./cli.log", "INFO")
```
Comment (kshlm) [inline cli/cmd/root.go]: This should be configurable from now itself.

File: cli/main.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/gluster/glusterd2/cli/cmd"
+)
+
+func main() {
+	// Migrate old format Args into new Format. Modifies os.Args()
```
Comment (kshlm) [inline cli/main.go]: Is there a way to get the log initialized here before we do anything else?

Right now, it's being initialized during package initialization of `cmd`, which technically happens before this point, but it doesn't allow for much configurability.

File: cli/cmd/root.go (hunk 2 above)
Comment (prashanthpai) [inline cli/cmd/root.go]: Also, please check the return value for error.

File: cli/cmd/volume.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+		volname := cmd.Flags().Args()[0]
+		fmt.Println("START:", volname)
+	},
+}
+
+var volumeStopCmd = &cobra.Command{
+	Use:   "stop",
+	Short: helpVolumeStopCmd,
+	Run: func(cmd *cobra.Command, args []string) {
+		validateNArgs(cmd, 1, 1)
```
Comment (prashanthpai) [inline cli/cmd/volume.go]: I so miss Python's decorators :)

File: cli/cmd/volume.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
+package cmd
+
+import (
+	"fmt"
+	"github.com/spf13/cobra"
+)
+
+const (
+	helpVolumeCmd       = "Gluster Volume Management"
```
Comment (prashanthpai) [inline cli/cmd/volume.go]: As purely a personal preference, unless these were used at multiple places, I'd place them in the declaration of `cobra.Command` itself.

File: scripts/build-cli.sh [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  OUTDIR=$1
+fi
+
+GOBUILD_TAGS="novirt noaugeas "
+VERSION=$($(dirname $0)/pkg-version --full)
+REPO_PATH="github.com/gluster/glusterd2"
+GIT_SHA=`git rev-parse --short HEAD || echo "undefined"`
+LDFLAGS="-X ${REPO_PATH}/version.GlusterdVersion=$VERSION -X ${REPO_PATH}/version.GitSHA=$GIT_SHA"
+LDFLAGS+=" -B 0x$(head -c20 /dev/urandom | od -An -tx1 | tr -d ' \n')"
+BIN=gluster
```
Comment (aravindavk) [inline scripts/build-cli.sh]: Makes sense. will change it to `glustercli`

File: cli/cmd/volume.go (hunk 5 above)
Comment (aravindavk) [inline cli/cmd/volume.go]: Placed it separately only to facilitate easy translation of these messages in future.

File: cli/main.go (hunk 3 above)
Comment (aravindavk) [inline cli/main.go]: I moved initialization code before cmd initialization. Not yet made configurable. 

With the help of `PersistentPreRun` option in `cobra` I can initialize log there, but it is not possible to initialize log file before parsing the args. (Logger will not be available during Args Migrate). I need to write custom parser to parse `--log-file` and `--log-level` outside the `cobra.Command` initialization.

I will address this with separate patch, as of now added constants for the same.

I will address this in separate patch.

**Pull request merged**